### PR TITLE
conf(env) Kuiper can be configured via env variables

### PR DESCRIPTION
**Pull request by rwadowski**
@ngjaying it turns out that there is a ready library to use for this purpose

### CODE REVIEW

**Review by ngjaying — COMMENTED** [review]
Great work. Just several concerns from me:
1. We actually only need a small protion of the features of viper. Could you check how much does the binary size increase after adding this dependency.
2. Besides kuiper.yaml, we also have other yaml configs in etc/sources and maybe other places. We may need to support all of them as well. It will be better to eliminate the direct dependenct to gopkg.in/yaml.v3 after all changes.

**Comment by rwadowski** [comment]
> Great work. Just several concerns from me:
> 
> 1. We actually only need a small protion of the features of viper. Could you check how much does the binary size increase after adding this dependency.

`-rwxrwxr-x 1 rwadowski rwadowski 23841760 Aug 23 10:35 no_viper`
`-rwxrwxr-x 1 rwadowski rwadowski 25699552 Aug 23 10:34 with_viper`

It is more less 7% increase of memory size

**Comment by rwadowski** [comment]
@ngjaying 
> 2\. Besides kuiper.yaml, we also have other yaml configs in etc/sources and maybe other places. We may need to support all of them as well. It will be better to eliminate the direct dependenct to gopkg.in/yaml.v3 after all changes.
I guess you are writing for example about conf_util.go which seems to do similar thing

**Comment by ngjaying** [comment]
> @ngjaying
> 
> > 2. Besides kuiper.yaml, we also have other yaml configs in etc/sources and maybe other places. We may need to support all of them as well. It will be better to eliminate the direct dependenct to gopkg.in/yaml.v3 after all changes.
> > I guess you are writing for example about conf_util.go which seems to do similar thing

Not just conf_util.go. There are many other places like internal/topo/node/node.go that is using "gopkg.in/yaml.v3" to parse config like etc/sources/edgex.yaml.

**Comment by rwadowski** [comment]
> > @ngjaying
> > > 1. Besides kuiper.yaml, we also have other yaml configs in etc/sources and maybe other places. We may need to support all of them as well. It will be better to eliminate the direct dependenct to gopkg.in/yaml.v3 after all changes.
> > >    I guess you are writing for example about conf_util.go which seems to do similar thing
> 
> Not just conf_util.go. There are many other places like internal/topo/node/node.go that is using "gopkg.in/yaml.v3" to parse config like etc/sources/edgex.yaml.

I guess it will be faster for me to remove dependency and see what will crash

**Comment by ngjaying** [comment]
> > Great work. Just several concerns from me:
> > 
> > 1. We actually only need a small protion of the features of viper. Could you check how much does the binary size increase after adding this dependency.
> 
> `-rwxrwxr-x 1 rwadowski rwadowski 23841760 Aug 23 10:35 no_viper`
> `-rwxrwxr-x 1 rwadowski rwadowski 25699552 Aug 23 10:34 with_viper`
> 
> It is more less 7% increase of memory size

It looks like an obvious increase for more than 1MB. I don't know if increase can be much smaller if we can remove all yaml dependencies.

**Comment by rwadowski** [comment]
@ngjaying what's more it looks like viper depends on the gopkg.in/yaml.v3

**Comment by rwadowski** [comment]
> It looks like an obvious increase for more than 1MB. I don't know if increase can be much smaller if we can remove all yaml dependencies.

It seems that not all yaml dependencies will be removed as I mentioned in other post due to vipers dependency on the yaml ones.

**Comment by ngjaying** [comment]
> > It looks like an obvious increase for more than 1MB. I don't know if increase can be much smaller if we can remove all yaml dependencies.
> 
> @ngjaying It seems that not all yaml dependencies will be removed as I mentioned in other post due to vipers dependency on the yaml ones. So the question whether increase of footprint will be acceptable at all ?

If not very hard to implement by ourselves, I would prefer to not introduce vipers dependency.

**Comment by rwadowski** [comment]
> > > It looks like an obvious increase for more than 1MB. I don't know if increase can be much smaller if we can remove all yaml dependencies.
> > 
> > 
> > @ngjaying It seems that not all yaml dependencies will be removed as I mentioned in other post due to vipers dependency on the yaml ones. So the question whether increase of footprint will be acceptable at all ?
> 
> If not very hard to implement by ourselves, I would prefer to not introduce vipers dependency.

Well I am far from being a golang expert but I guess it is possible to iterate over config struct and look for environment variables built on path in the tree - this would be a relatively generic way of loading envs. Is it something you think about ?

**Comment by ngjaying** [comment]
> > > > It looks like an obvious increase for more than 1MB. I don't know if increase can be much smaller if we can remove all yaml dependencies.
> > > 
> > > 
> > > @ngjaying It seems that not all yaml dependencies will be removed as I mentioned in other post due to vipers dependency on the yaml ones. So the question whether increase of footprint will be acceptable at all ?
> > 
> > 
> > If not very hard to implement by ourselves, I would prefer to not introduce vipers dependency.
> 
> Well I am far from being a golang expert but I guess it is possible to iterate over config struct and look for environment variables built on path in the tree - this would be a relatively generic way of loading envs. Is it something you think about ?

It could be a refactor of conf_util.go, but instead of write to file, just rewrite the conf values in memory.

**Comment by rwadowski** [comment]
> > > > > It looks like an obvious increase for more than 1MB. I don't know if increase can be much smaller if we can remove all yaml dependencies.
> > > > 
> > > > 
> > > > @ngjaying It seems that not all yaml dependencies will be removed as I mentioned in other post due to vipers dependency on the yaml ones. So the question whether increase of footprint will be acceptable at all ?
> > > 
> > > 
> > > If not very hard to implement by ourselves, I would prefer to not introduce vipers dependency.
> > 
> > 
> > Well I am far from being a golang expert but I guess it is possible to iterate over config struct and look for environment variables built on path in the tree - this would be a relatively generic way of loading envs. Is it something you think about ?
> 
> It could be a refactor of conf_util.go, but instead of write to file, just rewrite the conf values in memory.

@ngjaying will refactor of the conf_util.go will apply for normal app runtime ? As I see it is in /deploy/docker package which probably applies for docker. I think we need something that will apply env variable functionality to any type of deployment. I am not sure how this precisely works though ...

**Comment by ngjaying** [comment]
> @ngjaying will refactor of the conf_util.go will apply for normal app runtime ? As I see it is in /deploy/docker package which probably applies for docker. I think we need something that will apply env variable functionality to any type of deployment. I am not sure how this precisely works though ...

Yes, exactly
1. Merge conf_util to the "app runtime" instead of outside of the main package
2. Refactor it to onveride the env variable to memory instead of changing the conf file
3. Maybe a more graceful way to deal with the config key case as the env are upper case but the conf key varies.

**Comment by rwadowski** [comment]
> Yes, exactly
> 
> 1. Merge conf_util to the "app runtime" instead of outside of the main package
> 2. Refactor it to onveride the env variable to memory instead of changing the conf file
> 3. Maybe a more graceful way to deal with the config key case as the env are upper case but the conf key varies.

So keeping simple hardcoded map of env keys is fine ? We will have to add new keys there if there will be a new feature but after all new config will have to be added as well so env handling can be done in this case too.

**Comment by ngjaying** [comment]
> > Yes, exactly
> > 
> > 1. Merge conf_util to the "app runtime" instead of outside of the main package
> > 2. Refactor it to onveride the env variable to memory instead of changing the conf file
> > 3. Maybe a more graceful way to deal with the config key case as the env are upper case but the conf key varies.
> 
> So keeping simple hardcoded map of env keys is fine ? We will have to add new keys there if there will be a new feature but after all new config will have to be added as well so env handling can be done in this case too.

Do you mean `file_keys_map`? That is not hardcoded map of all keys. It is actually a map of upper case key to the camel case which cannot be calculated. For keys like `debug` are not in the map but can still be used as they are single word and can be deduced by just turn the env to lowercase. And this is needed to be refactored as my point 3.

**Comment by rwadowski** [comment]
@ngjaying shouldn't we take assumption that env variable coresponds to name in config for eample
```
basic:
   debug: true
```
should have env variable like `KUIPER_BASIC_DEBUG` ?
This will give us a very easy naming convention for env variables.

**Comment by ngjaying** [comment]
> @ngjaying shouldn't we take assumption that env variable coresponds to name in config for eample
> 
> ```
> basic:
>    debug: true
> ```
> 
> should have env variable like `KUIPER_BASIC_DEBUG` ?
> This will give us a very easy naming convention for env variables.

We do have the convention. This is the current convention `KUIPER__BASIC__DEBUG` which will map to kuper.yaml/basic/debug. And this does NOT need any hardcoded map. However, `KUIPER__BASIC__CONSOLELOG` currently needs a map, as we cannot know which alphabet need to be uppercase as the config key is consoleLog. The problem may be that we should support case insensitive config key.

**Comment by rwadowski** [comment]
> > @ngjaying shouldn't we take assumption that env variable coresponds to name in config for eample
> > ```
> > basic:
> >    debug: true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > should have env variable like `KUIPER_BASIC_DEBUG` ?
> > This will give us a very easy naming convention for env variables.
> 
> We do have the convention. This is the current convention `KUIPER__BASIC__DEBUG` which will map to kuper.yaml/basic/debug. And this does NOT need any hardcoded map. However, `KUIPER__BASIC__CONSOLELOG` currently needs a map, as we cannot know which alphabet need to be uppercase as the config key is consoleLog. The problem may be that we should support case insensitive config key.

Yeah, I was going into this direction that make all upper case envs like in linuks. As for now I extracted single file parsing into the internal/conf/load.go

**Comment by rwadowski** [comment]
I am looking for places in the code where edgex.yaml, random.yaml, zmq.yaml, httppull.yaml, client.yaml are loaded

**Comment by rwadowski** [comment]
@ngjaying I have applied all changes with difference of the env naming. Can you look at it ? Without mapping code is simpler I think

**Comment by rwadowski** [comment]
@ngjaying another thing about config stuff
instead of 
```
sink:
  # The cache persistence threshold size. If the message in sink cache is larger than 10, then it triggers persistence. If you find
  # the remote system is slow to response, or sink throughput is small, then it's recommend to increase below 2 configurations.
  # More memory is required with the increase of below 2 configurations.
  # If the message count reaches below value, then it triggers persistence.
  cacheThreshold: 10
  # The message persistence is triggered by a ticker, and cacheTriggerCount is for using configure the count to trigger the persistence procedure
  # regardless if the message number reaches cacheThreshold or not. This is to prevent the data won't be saved as the cache never pass the threshold.
  cacheTriggerCount: 15
```
shouldn't be 
```
sink:
  cache:
    # The cache persistence threshold size. If the message in sink cache is larger than 10, then it triggers persistence. If you find
    # the remote system is slow to response, or sink throughput is small, then it's recommend to increase below 2 configurations.
    # More memory is required with the increase of below 2 configurations.
    # If the message count reaches below value, then it triggers persistence.
    threshold: 10
    # The message persistence is triggered by a ticker, and cacheTriggerCount is for using configure the count to trigger the 
    #persistence procedure
    # regardless if the message number reaches cacheThreshold or not. This is to prevent the data won't be saved as the cache   
    # never pass the threshold.
  trigger:
    count: 15
```

**Review by ngjaying — CHANGES_REQUESTED** [review]

**Comment by ngjaying** [comment]
> @ngjaying another thing about config stuff
> instead of
> 
> ```
> sink:
>   # The cache persistence threshold size. If the message in sink cache is larger than 10, then it triggers persistence. If you find
>   # the remote system is slow to response, or sink throughput is small, then it's recommend to increase below 2 configurations.
>   # More memory is required with the increase of below 2 configurations.
>   # If the message count reaches below value, then it triggers persistence.
>   cacheThreshold: 10
>   # The message persistence is triggered by a ticker, and cacheTriggerCount is for using configure the count to trigger the persistence procedure
>   # regardless if the message number reaches cacheThreshold or not. This is to prevent the data won't be saved as the cache never pass the threshold.
>   cacheTriggerCount: 15
> ```
> 
> shouldn't be
> 
> ```
> sink:
>   cache:
>     # The cache persistence threshold size. If the message in sink cache is larger than 10, then it triggers persistence. If you find
>     # the remote system is slow to response, or sink throughput is small, then it's recommend to increase below 2 configurations.
>     # More memory is required with the increase of below 2 configurations.
>     # If the message count reaches below value, then it triggers persistence.
>     threshold: 10
>     # The message persistence is triggered by a ticker, and cacheTriggerCount is for using configure the count to trigger the 
>     #persistence procedure
>     # regardless if the message number reaches cacheThreshold or not. This is to prevent the data won't be saved as the cache   
>     # never pass the threshold.
>     trigger:
>       count: 15
> ```
> 
> ?

There are not many configurations under sink, so we just flat them to avoid introducing too many hierarchies.

File: internal/conf/load.go
Comment (rwadowski) [reply internal/conf/load.go:60]: @ngjaying I am not sure how to make it better, I need to convert map[string]interface{} to struct / interface

Comment (rwadowski) [reply internal/conf/load.go:60]: @ngjaying in case of camelCase. I have not handled it properly - I will need to add mapping but I have feeling that proper hierarchy in they yaml config would solve the problem

Comment (ngjaying) [reply internal/conf/load.go:60]: ah ok. I think current approach is ok. It is just an one time action. Does this also solve the problem of lowercase -> camelCase?

Comment (rwadowski) [reply internal/conf/load.go:60]: nope, I am about commit fix about camel case handling,I have changed Unmatshalling/Marshalling to sth that looks better at a cost of additional dependency.

Comment (rwadowski) [reply internal/conf/load.go]: I have commited it right now, I have feeling that it looks better right now

File: extensions/go.mod
Comment (rwadowski) [reply extensions/go.mod]: Fixed

**Comment by rwadowski** [comment]
@ngjaying what check logs step in run_fvt_tests does ?

**Comment by rwadowski** [comment]
> @rwadowski I would like to get rid of the mapping. I think it is possible like adding a json tag in addition to yaml tag in the config strutct fields. Let the json tag be case insensitive like all lowercase and try to leverage it to do the decode.

@ngjaying can you elaborate more ? I don't know golang ecosystem well. I am trying to find something how to make it feasible but I can't see anything that would be handy.

File: internal/conf/load.go
Comment (ngjaying) [inline internal/conf/load.go:109]: I suppose the keys in the configMap are camelCase as they specified in the yaml.  Is the yaml tag in the config struct like  ```ConsoleLog   bool   `yaml:"consoleLog"` ``` is still used in the decode of line 116? For envs, if we let the key to be all lowercase and just set into the configMap, is it possible to let the decode map to the correct case. Maybe it is possible to add a new tag in the config struct field like ```ConsoleLog   bool   `yaml:"consoleLog",json:"consolelog"` ```  in which the json tag name is all lowercase.

Comment (rwadowski) [reply internal/conf/load.go:51]: Is something I committed fine ? I leveraged [this](https://pkg.go.dev/github.com/mitchellh/mapstructure#hdr-Field_Tags)

**Comment by ngjaying** [comment]
We are releasing 1.3.1 today. After then, we will merge back dev-1.4.0 to master and try to merge this. Thank you!

**Comment by rwadowski** [comment]
> We are releasing 1.3.1 today. After then, we will merge back dev-1.4.0 to master and try to merge this. Thank you!

I assume that review can be closed ?

**Comment by ngjaying** [comment]
> > We are releasing 1.3.1 today. After then, we will merge back dev-1.4.0 to master and try to merge this. Thank you!
> 
> I assume that review can be closed ?

We can keep this. Just rebase it on the new master later.

**Comment by ngjaying** [comment]
Force push to this after rebasing the new master. Will do some smoke testing and merge this if all are fine. Thanks.

File: cmd/kuiper/main.go
Comment (ngjaying) [inline cmd/kuiper/main.go:67]: This is wrong. Here we should read clientYaml: client.yaml but the LoadConfig read from kuiper.yaml

File: internal/conf/load.go
Comment (ngjaying) [inline internal/conf/load.go:56]: I suppose here is the case of source config reading. Even read into map, we still need to deal with case. For example, edgex.yaml has a key "messageType" in yaml and after reading the map key is also camel case "messageType"; However, the env var will be parsed to "messagetype". How do we deal with this right now?

**Review by ngjaying — CHANGES_REQUESTED** [review]

File: internal/conf/load.go
Comment (ngjaying) [reply internal/conf/load.go:56]: maybe we can get the key case from its json file, for example edgex.json has a edgex.json file with properties section that listed all the available properties.

File: cmd/kuiper/main.go
Comment (rwadowski) [reply cmd/kuiper/main.go]: Fixed

File: internal/conf/load.go
Comment (rwadowski) [reply internal/conf/load.go:56]: @ngjaying I disabled normalization for map type config due to one of failing tests. Maybe we should fix the test instead of creating an edge case for the map type ?

Comment (ngjaying) [reply internal/conf/load.go:56]: There is actually a big difference here. For struct, we can implicitly define the case mapping by the yaml tag. But for map, there is no tag. How do you know what an env var MESSAGETYPE should be converted to?

Comment (rwadowski) [reply internal/conf/load.go:56]: I think it would good to make the same behavior - make everything case insensitive. Somehow it does not convince me that for yaml we should be case insensitive while for map case sensitive.

Comment (ngjaying) [reply internal/conf/load.go:56]: yaml is case sensitive now as we read it from `yaml.Unmarshal(b, &output)`. We did not want to change this behavior. When the output is a struct, the case sensitive yaml key are map to the struct field with the help of yaml tag of the struct. Othwise, if the outpus is a map, the case sensitive yaml key directly become the map key. For example, yaml key `messageType` will become `MessageType` in a struct and become `messageType` in a map. 

The problem is that the env is case insensitive and you normalize key to handle it. If the output is a struct, we can normalize the key and changes back to the case sensitive struct key with the help of mapstructure decode. But for map, if we normalize the key, we cannot revert back. We seems to miss the key case information. And that's my concern.

Comment (ngjaying) [reply internal/conf/load.go:56]: Moreover, the struct always define all available keys(fields), but map we only read from yaml. If the yaml does not define all keys, we have no way to know if a key from ENV is valid or not. I think the root cause is the missing information of property meta data for source configure. We can actually read that meta data from an optional json file like edgex.json.

Comment (rwadowski) [reply internal/conf/load.go:56]: > Moreover, the struct always define all available keys(fields), but map we only read from yaml. If the yaml does not define all keys, we have no way to know if a key from ENV is valid or not. I think the root cause is the missing information of property meta data for source configure. We can actually read that meta data from an optional json file like edgex.json.

You mean add such functionality - json to map[string]interface{} ?

Comment (ngjaying) [reply internal/conf/load.go:56]: No. There is an optional json file for each plugin. For example, etc/sources/edgex.json defines the meta data of edgex source.  Inside it, there is a section properties which define all the available properties with the correct case. When parsing yaml to map we can use the properties information to validate the keys. When convert ENV to conf, we can also use the properties information to map the key from upper case to the correct case.

Comment (rwadowski) [reply internal/conf/load.go:56]: It is an array - how does this refer to config ?

Comment (ngjaying) [reply internal/conf/load.go:56]: Let me take edgex as an example. From etc/sources/edgex.json, we can get a list of correct case properties(keys) like ["protocol", "messageType"...]. In internal/topo/node/node.go line 182, you parse yaml to a map. At there, we can loop the map key and validate against the key array of json. Then in this file, when handling env, we can still normalize the key and when decoding back to map, use the key array of json to restore the case of the keys.

Comment (rwadowski) [reply internal/conf/load.go:56]: ok I think I finally understood what you are writing about, where I am supposed to look for these files. One file at particular ocassion will suffice or particular yaml requires particular json ?

Comment (ngjaying) [reply internal/conf/load.go:56]: Great! Each yaml should have a json file alongside for all built-in sources. It is optional for extended plugin, so if json is not found, we can just skip this check.

Comment (rwadowski) [reply internal/conf/load.go:56]: Awesome, are these json locations the same as yaml counterparts or they have specific location ?

Comment (ngjaying) [reply internal/conf/load.go:56]: yes, they are at the same location.

Comment (rwadowski) [reply internal/conf/load.go]: hi, I made the changes but I have problems [here](https://github.com/lf-edge/ekuiper/runs/3549255802?check_suite_focus=true), can you help me ?

Comment (ngjaying) [reply internal/conf/load.go]: Hi Robert, that should be a network problem of the tdengine client download. We are aware of it and just check with the tdengine team. Let's just leave it there.

Comment (rwadowski) [reply internal/conf/load.go]: What about changes, is this you were writing about ?

Comment (ngjaying) [reply internal/conf/load.go]: Yes, thanks

**Review by ngjaying — DISMISSED** [review]

**Comment by ngjaying** [comment]
I'll need to rebase and merge it.

File: internal/conf/load.go
Comment (ngjaying) [inline internal/conf/load.go:74]: The Prefix should not be a constant. It should be a variable. When parsing a source config, the prefix is the source name. For example, EDGEX__DEFAULT__PROTOCOL.

**Review by ngjaying — CHANGES_REQUESTED** [review]
@rwadowski I have merged this and add a log to help debugging if env has applied successfully. I also find another problem when trying locally, I have added it to the review. Notice that, I have force push my changes to your branch. When you continue working with this, please fetch and reset to the latest. `git fetch origin` `git reset --hard ISSUE-962`

File: internal/conf/load_test.go
Comment (ngjaying) [inline internal/conf/load_test.go:45]: Here the key prefix should be the same as the yaml name. HTTPPULL_DEFAULT_BODYTYPE

File: internal/conf/load.go
Comment (rwadowski) [reply internal/conf/load.go:74]: ok so how can I determine prefix ?

File: internal/conf/load_test.go
Comment (rwadowski) [reply internal/conf/load_test.go]: I have pushed fixes

File: internal/conf/load.go
Comment (ngjaying) [inline internal/conf/load.go:48]: Remove this

**Review by ngjaying — CHANGES_REQUESTED** [review]
It looks good now except one unnecessary print. Also could you use rebase to get rid of the merge commits to keep a linear history.

File: internal/conf/load.go
Comment (rwadowski) [reply internal/conf/load.go]: After applying fixes this test started to fail - [here](https://github.com/lf-edge/ekuiper/actions/runs/1216670656) which is pretty surprising

Comment (ngjaying) [reply internal/conf/load.go]: It is a timeout, usually rerun would solve it.

**Review by ngjaying — APPROVED** [review]
Merging

**Pull request merged**
